extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java

```diff
@@ -0,0 +1,88 @@
+package io.quarkus.cache.runtime;
+
+import static javax.interceptor.Interceptor.Priority.PLATFORM_BEFORE;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.annotation.Priority;
+import javax.inject.Inject;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.cache.CacheInvalidate;
+import io.quarkus.cache.CacheInvalidateAll;
+import io.quarkus.cache.CacheKey;
+import io.quarkus.cache.CacheResult;
+import io.quarkus.cache.runtime.caffeine.CaffeineCache;
+
+@CacheInterceptorBinding
+@Interceptor
+@Priority(PLATFORM_BEFORE)
+public class CacheInterceptor {
+
+    private static final Logger LOGGER = Logger.getLogger(CacheInterceptor.class);
+
+    @Inject
+    CacheRepository cacheRepository;
+
+    @AroundInvoke
+    public Object intercept(InvocationContext context) throws Exception {
+        Method method = context.getMethod();
+
+        for (CacheInvalidateAll cacheInvalidateAll : method.getDeclaredAnnotationsByType(CacheInvalidateAll.class)) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debugf("Invalidating all entries from cache [%s]", cacheInvalidateAll.cacheName());
+            }
+            CaffeineCache cache = cacheRepository.getCache(cacheInvalidateAll.cacheName());
+            cache.invalidateAll();
+        }
+
+        for (CacheInvalidate cacheInvalidate : method.getDeclaredAnnotationsByType(CacheInvalidate.class)) {
+            Object key = getCacheKey(context, cacheInvalidate.cacheName());
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debugf("Invalidating entry with key [%s] from cache [%s]", key, cacheInvalidate.cacheName());
+            }
+            CaffeineCache cache = cacheRepository.getCache(cacheInvalidate.cacheName());
+            cache.invalidate(key);
+        }
+
+        CacheResult cacheResult = method.getDeclaredAnnotation(CacheResult.class);
+        if (cacheResult != null) {
+            Object key = getCacheKey(context, cacheResult.cacheName());
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debugf("Loading entry with key [%s] from cache [%s]", key, cacheResult.cacheName());
+            }
+            CaffeineCache cache = cacheRepository.getCache(cacheResult.cacheName());
+            return cache.get(key, () -> context.proceed(), cacheResult.lockTimeout());
+        }
```

Comment:
I wonder how efficient this code will be considering it will call `declaredAnnotations()` (which is a synchronized method) 3 times.

Reply:
The `synchronized` block should only be called once since the declared annotations will already be parsed when the 2nd and 3rd calls happen, but that's still a very interesting question. Thanks for taking a look at this!

Reply:
FYI, I tried to avoid using reflection and to do things at build time using annotations transformation originally, but there's a limitation in the current Quarkus annotation transformation logic that blocked me (I already had a discussion with @mkouba about it). I'd be happy to remove all reflection calls if that's possible :)

Reply:
So I'm a bit disappointed in the sense that witht he build time knowledge, we should be able to add an interceptor with the right state. We should know the annotations effectively present and even the cache name state and it should be burnt into the interceptor instance. I don't know how much of this is doable with the CDI model but that would be a nice improvement. Is that the conversation you had with @mkouba @gwenneg ?

Reply:
But as a temp measure, it looks like the synchronize call will only happen the first time the service class is requested its annotations. We could do this in a static_init block to do it at "startup time" if we feel its a problem to pay the cost lazily.

Reply:
I talked with @mkouba about the following annotation that I tried to introduce through transformation at build time:
```
@CacheInterceptorBinding(
    cacheInvalidateAll = { @CacheInvalidateAll(cacheName = "foo"), @CacheInvalidateAll(cacheName = "bar") },
    cacheInvalidate = { @CacheInvalidate(cacheName = "foo", @CacheInvalidate(cacheName = "bar") },
    cacheResult = @CacheResult(cacheName = "foo")
)
```
With such an annotation, most of the reflection calls wouldn't be required anymore (there would still be some reflection involved to identify the `@CacheKey`-annotated parameters).

Reply:
Maybe handling everything in a single interceptor is not a reasonnable idea. Another solution would be to keep one interceptor by caching operation and change the priorities to make sure the execution order is always the same.

Reply:
> it looks like the synchronize call will only happen the first time the service class is requested its annotations

That's not what I see as AFAICS `declaredAnnotations()` will always be called when calling `getDeclaredAnnotationsByType()`. It will be very fast after the first invocation but you still have the synchronized part. Unless I miss something obvious?